File: week6/sentimental-credit/credit.py
# from cs50 import get_string


def main():

    number = input("Number: ")

    if not checksum(number):
        print("INVALID")
        return

    first_two = f"{number[0]}{number[1]}"

    if first_two in ['34', '37'] and len(number) == 15:
        print("AMEX")
    elif int(first_two) in list(range(51, 56)) and len(number) == 16:
        print("MASTERCARD")
    elif first_two.startswith('4') and len(number) in [13,16]:
        print("VISA")
    else:
        print("INVALID")


def checksum(number):

    total = 0

    for i in range(len(number)):
        last_num = int(number[-1])
        number = number[:-1]
        if i % 2 == 0:
            total += last_num
            continue

        else:
            last_num *= 2

        if last_num >= 10:
            last_num = str(last_num)
            total += int(last_num[0]) + int(last_num[1])
            continue

        total += last_num

    return not total % 10

if __name__ == "__main__":
    main()

File: week1/Pset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

bool checksum(long num);

int count;

int main(void)
{
    long number = get_long("Number: ");

    if (!checksum(number))
    {
        printf("INVALID\n");
        return 0;
    }

    long n = number;

    int length = floor(log10(labs(number))) + 1;

    while (n >= 100)
    {
        n = number /= 10;
    }

    if ((n == 34 || n == 37) && length == 15)
    {
        printf("AMEX\n");
        return 0;
    }
    if ((n >= 51 && n <= 55) && length == 16)
    {
        printf("MASTERCARD\n");
        return 0;
    }
    if ((n /= 10) == 4 && (length == 13 || length == 16))
    {
        printf("VISA\n");
        return 0;
    }
    printf("INVALID\n");
    return 0;
}

bool checksum(long num)
{
    int total = 0;

    for (int i = 0; num > 0; i++)
    {
        int tmp = num % 10;
        num /= 10;
        if (i % 2 == 0)
        {
            total += tmp;
            continue;
        }
        tmp *= 2;
        if (tmp >= 10)
        {
            int a = tmp % 10;
            tmp /= 10;
            total += (tmp + a);
            continue;
        }
        total += tmp;
    }

    if (total % 10 == 0)
    {
        return true;
    }
    return false;
}
